### DISCUSSION TITLE
Another Seplos BMS 3.0

### DISCUSSION THREAD

**Discussion by Goaheadz** [syssi/esphome-seplos-bms#98] [Q&A]
I am not being able to connect to BMS using any of the example codes.

Using d1 mini with RS485 adapter.
TX/RX - GPIO5 and GPIO4

The RJ45 cable is ok, cause is the same that I use with another USB-TTL and same RS485 board when i connect computer to the seplos bmsstudio.

![image](https://github.com/syssi/esphome-seplos-bms/assets/13796554/add621a0-0054-4989-bc4a-c2178668dba6)

I also swap TX with RX between ESP8266 and RS485 board, and result is the same...no response from BMS.

What else can I try and check?

Thx

**Comment by syssi** [comment]
1. Please double check the baudrate (9600 vs 19200).
2. Could you try to capture some traffic using the Seplos BMSStudio app to make sure we are using the correct protocol/frame version?

**Reply by Goaheadz** [reply]
Code is on 19200, just like when I connect with BMSStudio.
What free app can I use for traffic capture in com8?

**Reply by syssi** [reply]
If remember correctly the BMStudio app has a monitor view or RAW traffic export button or something like that. That would be sufficient.

**Reply by Goaheadz** [reply]
dammmm
I killed another TTL-RS485 board. :(
Cannot connect to BMS with seplos SW. Tx side works (wakes up LCD) but RX led is always on.
Cheap aliexpress boards. Its the second one out of 3 that dies.
The working one is currently connect to HVAC, cannot use it.

Well, I will not be able to move forward in the next weeks. No way to connect throught RS485. 

Have to order new HW. Its going to take a while.

But thanks for your help @syssi

**Reply by Goaheadz** [reply]
I am not using GND pin from the RS485 side, so probably thats what is killing the TTL-RS485.

**Comment by Goaheadz** [comment]
Dont have a way to connect BMS with D1 mini, but I was able to connect bmstudio via USB-rs485
[Communications Records 20240604_1958.txt](https://github.com/user-attachments/files/15570190/Communications.Records.20240604_1958.txt)
Here are some comunication records.

**Reply by syssi** [reply]
I've extracted the important party. `>>>` indicates outgoing traffic and `<<<` is incoming traffic / the response:

```
# Acquire AllFunction SwitchParameters(Pack#0)
>>> 00 01 14 00 00 50 38 17
<<< 00 01 0A FF FF 3F 01 7F 2F 9F 0F 63 6F 54 A1

# Acquire AllInt Parameters(Pack#0)
>>> 00 04 13 00 00 6A 75 70
<<< 00 04 D4 00 04 00 10 15 2C 15 E0 15 18 16 80 12 C0 12 20 12 C0 10 E0 0D 48 0D AC 0D 48 0E 10 0C 1C 0B 54 0C 1C 0A 8C 07 D0 03 E8 01 F4 00 CB 00 CD 00 D2 00 64 01 2C 01 2C FF 35 FF 33 FF 2E 00 64 FE A2 01 2C FE D4 00 6A 02 58 00 05 0B B8 00 CD 00 0A 0D AC 0D 48 05 DC 03 E8 03 20 00 C8 00 1E 0C 81 0C 9F 0C 9F 0C D1 0A DD 0A BF 0A AB 0A 47 0C 9F 0C D1 0C D1 0D 03 0A C9 0A 47 0A AB 0A 15 0C 81 0C 9F 0C D1 0D 03 0A C9 0A AB 0A AB 0A 47 0D FD 0E 61 0D FD 0E F7 0B 0F 0A AB 0C 9F 0A AB 00 0A 0D 48 00 0F 00 05 03 C0 00 96 00 64 00 46 00 32 6D 60 6D 60 38 54 00 30 00 3C 00 F0 00 0A 00 07 00 00 00 0D 00 00 01 F4 01 2C 16 80 00 B4 FF 4C 00 05 00 08 49 E5

# Acquire Device Manufacture Info
>>> 00 04 17 00 00 33 B4 7A
<<< 00 04 66 58 5A 48 2D 45 6C 65 63 54 65 63 68 20 43 6F 2E 2C 4C 74 64 42 4D 53 31 36 53 32 30 30 41 2D 53 50 30 35 42 00 00 00 00 31 33 53 50 30 35 42 32 33 31 32 31 39 30 30 37 35 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 5E 5E

# Acquire Pack#0Parameters Serial battery number
>>> 00 04 13 01 00 01 65 5F
<<< 00 04 02 00 10 85 3C

# Acquire Pack#0Part Model
>>> 00 04 17 0A 00 0A 54 6A
<<< 00 04 14 42 4D 53 31 36 53 32 30 30 41 2D 53 50 30 35 42 00 00 00 00 67 9A

# Acquire PIA(Pack#0)
>>> 00 04 10 00 00 12 75 16
<<< 00 04 24 14 A6 00 00 38 54 6D 60 00 60 02 03 03 E8 00 04 0C E8 0B 9F 0C E8 0C E8 0B A4 0B 9B 00 00 00 B4 00 B4 03 E8 1E 8F

# Acquire PIA(Pack#0)
>>> 00 04 10 00 00 12 75 16
<<< 00 04 24 14 A6 00 00 38 54 6D 60 00 60 02 03 03 E8 00 04 0C E8 0B 9F 0C E8 0C E8 0B A4 0B 9B 00 00 00 B4 00 B4 03 E8 1E 8F

# Acquire PIB(Pack#0)
>>> 00 04 11 00 00 1A 75 2C
<<< 00 04 34 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0B A4 0B 9B 0B 9C 0B A3 0A AB 0A AB 0A AB 0A AB 0B C2 0B B0 7B FF

# Acquire PIB(Pack#0)
>>> 00 04 11 00 00 1A 75 2C
<<< 00 04 34 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0C E8 0B A4 0B 9B 0B 9C 0B A3 0A AB 0A AB 0A AB 0A AB 0B C2 0B B0 7B FF

# Acquire PIC(Pack#0)
>>> 00 01 12 00 00 90 38 CF
<<< 00 01 12 00 00 00 00 00 00 00 00 10 00 00 00 00 00 00 03 00 00 6A 24

# Acquire PIC(Pack#0)
>>> 00 01 12 00 00 90 38 CF
<<< 00 01 12 00 00 00 00 00 00 00 00 10 00 00 00 00 00 00 03 00 00 6A 24

# Set AllParameters(Pack#0)
>>> 00 10 13 00 00 6A D4 00 04 00 10 15 18 15 E0 15 18 16 80 12 C0 12 20 12 C0 10 E0 0D 48 0D AC 0D 48 0E 10 0C 1C 0B 54 0C 1C 0A 8C 07 D0 03 E8 01 F4 00 CB 00 CD 00 D2 00 64 01 2C 01 2C FF 35 FF 33 FF 2E 00 64 FE A2 01 2C FE D4 00 6A 02 58 00 05 0B B8 00 CD 00 0A 0D AC 0D 48 05 DC 03 E8 03 20 00 C8 00 1E 0C 81 0C 9F 0C 9F 0C D1 0A DD 0A BF 0A AB 0A 47 0C 9F 0C D1 0C D1 0D 03 0A C9 0A 47 0A AB 0A 15 0C 81 0C 9F 0C D1 0D 03 0A C9 0A AB 0A AB 0A 47 0D FD 0E 61 0D FD 0E F7 0B 0F 0A AB 0C 9F 0A AB 00 0A 0D 48 00 0F 00 05 03 C0 00 96 00 64 00 46 00 32 6D 60 6D 60 38 54 00 30 00 3C 00 F0 00 0A 00 07 00 00 00 0D 00 00 01 F4 01 2C 16 80 00 B4 FF 4C 00 05 00 08 68 C2
<<< 00 10 13 00 00 6A 45 73
```

**Comment by Goaheadz** [comment]
From the txtt file above:

Acquire Device Manufacture Info
00 04 17 00 00 33 B4 7A 

returns:
BMS response OK
00 04 66 58 5A 48 2D 45 6C 65 63 54 65 63 68 20 43 6F 2E 2C 4C 74 64 42 4D 53 31 36 53 32 30 30 41 2D 53 50 30 35 42 00 00 00 00 31 33 53 50 30 35 42 32 33 31 32 31 39 30 30 37 35 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 5E 5E 

It means: (converted from HEX):
fXZH-ElecTech Co.,LtdBMS16S200A-SP05B
13SP05B2312190075
^^

Dont know if you are making the right request with your script..but this is what I get from seplos SW connected to BMS.

**Reply by syssi** [reply]
```
00 04 17 00 00 33 B4 7A
 |  |  |  |  |  |  |  |
 |  |  |  |  |  |  |  crc
 |  |  |  |  |  |  crc
 |  |  |  |  |  number of registers to read low byte -> 0x33 = 51 * 2 bytes = 102 bytes
 |  |  |  |   number of registers to read high byte
 |  |  |  start address low byte -> 0x1700
 |  |  start address high byte
 |  function code 4 (Read Input Registers)
 device address
```

This is a valid Modbus request which retrieves data starting at register 0x1700 reading 51 registers a 2 bytes. This is the response:

```
00 04 66 58 5A 48 2D 45 6C 65 63 54 65 63 68 20 43 6F 2E 2C 4C 74 64 42 4D 53 31 36 53 32 30 30 41 2D 53 50 30 35 42 00 00 00 00 31 33 53 50 30 35 42 32 33 31 32 31 39 30 30 37 35 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 5E 5E
 |  |  | ^^^^^content of register 0x1700
 |  |  |
 |  |  data length
 |  function code
device address
```

This matches with the protocol description here: https://github.com/syssi/esphome-seplos-bms/blob/main/docs/XZH%20BMS%20Modbus-RTU%20Protocol.pdf

![docs](https://github.com/syssi/esphome-seplos-bms/assets/2735933/b0cd5716-1f8a-4e21-b970-6d53803e4378)

```
20+20+2+30+30 = 102 bytes
```

The first 20 bytes of the response is the `factory name`. Next 20 bytes the `device name`. 2 bytes the `firmware version` etc.

**Reply by syssi** [reply]
IMO this YAML (https://github.com/syssi/esphome-seplos-bms/blob/main/esp8266-seplos-v3-example.yaml) should be compatible with your BMS in general. There must be something wrong with your wiring or RS485 converter module.

**Reply by Goaheadz** [reply]
Finally some RS485->TTL boards arrived
Its working like a charm!

Thank you so much for all the time you put into this.

![image](https://github.com/syssi/esphome-seplos-bms/assets/13796554/cc2740b8-aa87-45b7-9439-396027bb22cd)

**Reply by syssi** [reply]
Could you provide your BMS model and version? I would like to extend the list of supported devices if needed.

**Reply by Goaheadz** [reply]
Here we go:

Seplos BMS V3.0 Type C
Model nº: B-48200-C 
![image](https://github.com/syssi/esphome-seplos-bms/assets/13796554/85ca48df-5982-4651-bf41-74d8e335348e)

**Reply by Goaheadz** [reply]
From de DevBMSStudio:

![image](https://github.com/syssi/esphome-seplos-bms/assets/13796554/fe067324-aa89-4652-8532-635e5d91b577)
